File: Java/medium/longestCommonSubsequence.java
import java.util.*;

class Solution {
    public static int longestCommonSubsequence(String text1, String text2) {
        int row = text1.length(), col = text2.length();
        int[][] dp = new int[row + 1][col + 1];
        for(int i = 1; i <= row; i++){
            for(int j = 1; j <= col; j++){
                if(text1.charAt(i - 1) == text2.charAt(j - 1)){
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                }
                else{
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        return dp[row][col];
    }

    // text1 = abcde   text2 = ace     Length of the longest common subsequence = 3
    // text1 = abc     text2 = abc     Length of the longest common subsequence = 3
    // text1 = abc     text2 = de      Length of the longest common subsequence = 0
    // text1 = bsbininm        text2 = jmjkbkjkv       Length of the longest common subsequence = 1

    public static void main(String[] args){
        List<String[]> testSet = Arrays.asList(
            new String[]{"abcde","ace"},
            new String[]{"abc","abc"},
            new String[]{"abc", "de"},
            new String[]{"bsbininm", "jmjkbkjkv"}
        );

        for(String[] test : testSet){
            System.out.println( "text1 = " + test[0] + "\ttext2 = " + test[1] + "\tLength of the longest common subsequence = " + longestCommonSubsequence(test[0],test[1]));
        }
    }
}

File: CPP/medium/longestCommonSubsequence.cpp
#include <string>
#include <vector>
#include <iostream>

int longestCommonSubsequence(std::string text1, std::string text2) {
    std::vector<std::vector<int>> dp(text1.size() + 1, std::vector<int>(text2.size() + 1, 0));
    int row = text1.size(), col = text2.size();
    for(int i = 0; i < row; i++){
        for(int j = 0; j < col; j++){
            if(text1[i] == text2[j]){
                dp[i + 1][j + 1] = dp[i][j] + 1;
            }
            else{
                dp[i + 1][j + 1] = std::max(dp[i][j+1], dp[i+1][j]);
            }
        }
    }
    
    return dp[row][col];
}

// text1 = abcde   text2 = ace     Length of the longest common subsequence = 3
// text1 = abc     text2 = abc     Length of the longest common subsequence = 3
// text1 = abc     text2 = de      Length of the longest common subsequence = 0
// text1 = bsbininm        text2 = jmjkbkjkv       Length of the longest common subsequence = 1

int main(){
    std::vector<std::vector<std::string>> testSet = {
        {"abcde","ace"},
        {"abc","abc"},
        {"abc", "de"},
        {"bsbininm", "jmjkbkjkv"}
    };

    for(auto test : testSet){
        std::cout << "text1 = " << test[0] << "\ttext2 = " << test[1] << "\tLength of the longest common subsequence = " << longestCommonSubsequence(test[0],test[1]) << std::endl;
    }
    return 0;
}